src/persist-client/src/lib.rs

```diff
@@ -569,6 +592,177 @@ mod tests {
         assert_eq!(listen.read_until(&6).await, expected_events);
     }
 
+    // Appends need to be contiguous for a shard, meaning the lower of an appended batch needs to
+    // match the current shard upper.
+    #[tokio::test]
+    async fn contiguous_append() {
+        mz_ore::test::init_logging_default("info");
+
+        let data = vec![
+            (("1".to_owned(), "one".to_owned()), 1, 1),
+            (("2".to_owned(), "two".to_owned()), 2, 1),
+            (("3".to_owned(), "three".to_owned()), 3, 1),
+            (("4".to_owned(), "vier".to_owned()), 4, 1),
+            (("5".to_owned(), "cinque".to_owned()), 5, 1),
+        ];
+
+        let id = ShardId::new();
+        let client = new_test_client().await;
+
+        let (mut write, read) = client.expect_open::<String, String, u64, i64>(id).await;
+
+        // Write a [0,3) batch.
+        write
+            .expect_append(&data[..2], write.upper().clone(), vec![3])
+            .await;
+        assert_eq!(write.upper(), &Antichain::from_elem(3));
+
+        // Appending a non-contiguous batch should fail.
+        // Write a [0,5) batch with the second writer.
+        let result = write
+            .append(
+                &data[4..5],
+                Antichain::from_elem(5),
+                Antichain::from_elem(6),
+            )
+            .await
+            .expect("external error");
+        assert_eq!(result, Ok(Err(Upper(Antichain::from_elem(3)))));
+
+        // Fixing the upper to make the write contiguous should make the append succeed.
+        write.expect_append(&data[2..5], vec![3], vec![6]).await;
+        assert_eq!(write.upper(), &Antichain::from_elem(6));
+
+        let mut snap = read.expect_snapshot(5).await;
+        assert_eq!(snap.read_all().await, all_ok(&data, 5));
+    }
+
+    // Per-writer appends can be non-contiguous (and don't need to start at [0]), as long as
```

Comment:
nit: IMO the "(and don't need to start at [0])" is slightly unclear but unnecessary. I'd just remove it. here and elsewhere